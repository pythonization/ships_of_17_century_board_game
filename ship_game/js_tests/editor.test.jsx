import {
    expect, test, vi,
    beforeAll, afterAll
} from 'vitest'
import renderer from 'react-test-renderer';

import Editor from '../static/ship_game/components/editor/editor'

beforeAll(async () => {

    // instead of createPortal, just render element inline
    // (to do not simulate all page HTML)
    vi.mock('react-dom', async () => ({
        ...(await vi.importActual(
            'react-dom'
        )),
        createPortal: vi.fn((element, node) => element),
    }))

    // instead of showing Modal - do nothing
    // (to do not simulate all page HTML)
    vi.mock('bootstrap', async () => ({
        ...(await vi.importActual(
            'bootstrap'
        )),
        Modal: vi.fn().mockImplementation(() => {
            return { show: vi.fn() };
        }),
    }))
    // +
    // can not do auto-mock like
    // vi.mock('bootstrap')
    // because vitest does not expect Modal will be called with new

    // we are not testing map here, only editor
    // do not render map inside, because it is too big
    vi.mock('../static/ship_game/components/map/map.jsx', () => ({
        default: vi.fn().mockReturnValue(
            'Map was here, but now it is mocked'
        )
    }))

})
afterAll(() => {
    vi.restoreAllMocks()
})

test('Render tests: editor (dialogs closed/opened)', () => {
    const test_r_instance = renderer.create(<Editor />)

    expect(
        test_r_instance.toJSON()
        // better use "asFragment()" instead of 'react-test-renderer'
    ).toMatchSnapshot();

    let settings_bt = test_r_instance.root.find(r_instance => (
        r_instance.type == 'button' &&
        r_instance.children[1].toLowerCase().includes('settings')
    ))
    let copy_bt = test_r_instance.root.find(r_instance => (
        r_instance.type == 'button' &&
        r_instance.children[1].toLowerCase().includes('copy code')
    ))

    // open settings dialog
    settings_bt.props.onClick();

    // #region make map smaller
    const input_rows = test_r_instance.root.find(r_instance => (
        r_instance.type == 'input' &&
        r_instance.props.id == "num_rows"
    ))
    const input_cols = test_r_instance.root.find(r_instance => (
        r_instance.type == 'input' &&
        r_instance.props.id == "num_cols"
    ))
    const input_ratio = test_r_instance.root.find(r_instance => (
        r_instance.type == 'input' &&
        r_instance.props.id == "keep_ratio"
    ))

    input_ratio.props.onChange({
        target: {
            checked: false,
        }
    })
    input_rows.props.onChange({
        target: {
            value: 3,
        }
    })
    input_cols.props.onChange({
        target: {
            value: 3,
        }
    })
    // #endregion

    expect(
        test_r_instance.toJSON()
        // better use "asFragment()" instead of 'react-test-renderer'
    ).toMatchSnapshot();

    // open copy map dialog
    copy_bt.props.onClick();

    expect(
        test_r_instance.toJSON()
        // better use "asFragment()" instead of 'react-test-renderer'
    ).toMatchSnapshot();
})
