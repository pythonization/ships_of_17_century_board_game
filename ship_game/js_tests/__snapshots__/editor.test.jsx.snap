// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render tests: editor (dialogs closed/opened) 1`] = `
[
  <div
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    Map was here, but now it is mocked
  </div>,
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-code"
    />
      Copy code of a map
  </button>,
  " ",
  <button
    className="btn btn-outline-secondary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-gears"
    />
      Settings
  </button>,
]
`;

exports[`Render tests: editor (dialogs closed/opened) 2`] = `
[
  <div
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    Map was here, but now it is mocked
  </div>,
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-code"
    />
      Copy code of a map
  </button>,
  " ",
  <button
    className="btn btn-outline-secondary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-gears"
    />
      Settings
  </button>,
  <div
    className="modal-header"
  >
    <h1
      className="modal-title"
      id="ship_modal_label"
    >
      Settings
    </h1>
    <button
      aria-label="Close"
      className="btn-close"
      data-bs-dismiss="modal"
      type="button"
    />
  </div>,
  <div
    className="modal-body"
  >
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        Map
      </div>
      <div
        className="col"
      >
        <select
          aria-label="Map"
          className="form-select"
          onChange={[Function]}
          value="g_finland_48x27"
        >
          <option
            value="g_finland_16x9"
          >
            Gulf of Finland [16:9 6P]
          </option>
          <option
            value="g_finland_48x27"
          >
            Gulf of Finland [48:27 8P]
          </option>
          <option
            value="g_draw"
          >
            Gulf (drawing) [48:27 8P]
          </option>
          <option
            value="i_draw"
          >
            Islands (drawing) [32:18 8P]
          </option>
          <option
            value="i_draw_square"
          >
            Islands (drawing, square) [25:25 8P]
          </option>
        </select>
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <label
          className="form-label"
          htmlFor="num_rows"
        >
          Rows
        </label>
      </div>
      <div
        className="col"
      >
        <input
          className="form-control"
          id="num_rows"
          min="1"
          onChange={[Function]}
          type="number"
          value={3}
        />
      </div>
    </div>
    <div
      className="row mt-2"
    >
      <div
        className="col"
      >
        <label
          className="form-label"
          htmlFor="num_cols"
        >
          Columns
        </label>
      </div>
      <div
        className="col"
      >
        <input
          className="form-control"
          id="num_cols"
          min="1"
          onChange={[Function]}
          type="number"
          value={3}
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col offset-6"
      >
        <div
          className="form-check form-switch mt-2"
        >
          <input
            checked={false}
            className="form-check-input"
            id="keep_ratio"
            onChange={[Function]}
            role="switch"
            type="checkbox"
          />
          <label
            className="form-check-label"
            htmlFor="keep_ratio"
          >
            Keep ratio
          </label>
        </div>
      </div>
    </div>
    <div
      className="alert alert-info mt-3"
      role="alert"
    >
      <p>
        <strong>
          Instructions:
        </strong>
      </p>
      <p>
        <kbd>
          Click
        </kbd>
         or 
        <kbd>
          Click + Drag
        </kbd>
         to toggle water.
      </p>
      <p>
        <kbd>
          Ctrl + Click
        </kbd>
         to toggle port.
      </p>
    </div>
  </div>,
]
`;

exports[`Render tests: editor (dialogs closed/opened) 3`] = `
[
  <div
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    Map was here, but now it is mocked
  </div>,
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-code"
    />
      Copy code of a map
  </button>,
  " ",
  <button
    className="btn btn-outline-secondary"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fa-solid fa-gears"
    />
      Settings
  </button>,
  <div
    className="modal-header"
  >
    <h1
      className="modal-title"
      id="ship_modal_label"
    >
      Map Code
    </h1>
    <button
      aria-label="Close"
      className="btn-close"
      data-bs-dismiss="modal"
      type="button"
    />
  </div>,
  <div
    className="modal-body"
  >
    <textarea
      className="w-100"
      readOnly={true}
      rows="10"
      value="[[\\"W\\",\\"W\\",\\"W\\"],[\\"W\\",\\"W\\",\\"W\\"],[\\"W\\",\\"W\\",\\"W\\"]]"
    />
  </div>,
]
`;
